From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ItsSmile <smulelp@gmail.com>
Date: Fri, 14 Nov 2025 03:52:48 +0100
Subject: [PATCH] no-self-metadata-sneaksprint-update


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 0b90b6f30ea09fb117281d5ddd2fc752d2c139b5..832ecfe1a43f1ce02882f08be8ca19ce00fb4350 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -1,6 +1,5 @@
 package net.minecraft.server;
 
-import com.google.common.collect.Sets;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
@@ -274,12 +273,36 @@ public class EntityTrackerEntry {
         DataWatcher datawatcher = this.tracker.getDataWatcher();
 
         if (datawatcher.a()) {
-            this.broadcastIncludingSelf(new PacketPlayOutEntityMetadata(this.tracker.getId(), datawatcher, false));
+
+            //QuickSpigot start - Dont sent sneak/sprint update metadata to self
+            final PacketPlayOutEntityMetadata metadata = new PacketPlayOutEntityMetadata(this.tracker.getId(), datawatcher, false);
+            final List<DataWatcher.WatchableObject> changed = metadata.b;
+            boolean updateSneakOrSprint = false;
+            for (DataWatcher.WatchableObject watchableObject : changed) {
+                if (watchableObject == null) {
+                    return;
+                }
+                Object value = watchableObject.b();
+                if (watchableObject.a() == 0 && value instanceof Byte) {
+                    byte flags = (Byte) value;
+                    boolean sneaking = (flags & 0x02) != 0; // bit 1
+                    boolean sprinting = (flags & 0x08) != 0; // bit 3
+                    updateSneakOrSprint = sneaking || sprinting;
+                }
+            }
+
+            if(updateSneakOrSprint){
+                this.broadcast(metadata);
+            }else {
+                this.broadcastIncludingSelf(metadata);
+            }
+            //QuickSpigot end
+
         }
 
         if (this.tracker instanceof EntityLiving) {
-            AttributeMapServer attributemapserver = (AttributeMapServer) ((EntityLiving) this.tracker).getAttributeMap();
-            Set set = attributemapserver.getAttributes();
+            final AttributeMapServer attributemapserver = (AttributeMapServer) ((EntityLiving) this.tracker).getAttributeMap();
+            final Set<AttributeInstance> set = attributemapserver.getAttributes();
 
             if (!set.isEmpty()) {
                 // CraftBukkit start - Send scaled max health
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java b/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java
index 62492650a0e8d480da1b3920b7fe61ed5119749d..5d77332e15847a46597cd80c2b07b9bc88d8b786 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java
@@ -6,7 +6,7 @@ import java.util.List;
 public class PacketPlayOutEntityMetadata implements Packet<PacketListenerPlayOut> {
 
     private int a;
-    private List<DataWatcher.WatchableObject> b;
+    public List<DataWatcher.WatchableObject> b;
 
     public PacketPlayOutEntityMetadata() {}
 
